truffle/contracts/CrdFunding.sol

```diff
@@ -0,0 +1,262 @@
+        Ranks[_name].active = false;
+        int256 index = findStringIndex(_name, ActiveRanks);
+        assert(index >= 0);
+        ActiveRanks[uint256(index)] = ActiveRanks[ActiveRanks.length - 1];
+        ActiveRanks.pop();
+        emit RanksActivation(_name, "DEACTIVATE");
+    }
+
+    function activateRank(string calldata _name) public onlyOwner {
+        require(
+            !isStringInTable(_name, ActiveRanks),
+            "This rank is already active"
+        );
+        ActiveRanks.push(_name);
+        Ranks[_name].active = true;
+        emit RanksActivation(_name, "ACTIVATE");
+    }
+
+    function setGoal(uint256 _newGoal) public onlyOwner {
+        goal = _newGoal;
+        emit goalModified(_newGoal);
+    }
+
+    function getGoal() public view returns (uint256) {
+        return goal;
+    }
+
+    function setDescription(string memory _newProjectDescription)
+        public
+        onlyOwner
+    {
+        projectDescription = _newProjectDescription;
+        emit projectDescriptionModified(_newProjectDescription);
+    }
+
+    function getDescription() public view returns (string memory) {
+        return projectDescription;
+    }
+
+    function setEndDate(uint256 _newEndDate) public onlyOwner {
+        endDate = _newEndDate;
+        emit endDateModified(_newEndDate);
+    }
+
+    function getEndDate() public view returns (uint256) {
+        return endDate;
+    }
+
+    function getTotal() public view returns (uint256) {
+        return Total;
+    }
+
+    function sendDonation(string memory _rankName, bool _claimReward)
+        public
+        payable
+    {
+        require(
+            Ranks[_rankName].active || areStringsEquals(_rankName, "_noRank"),
+            "No rank active with this name"
+        );
+        require(block.timestamp < endDate, "Funding ended");
+        require(
+            msg.value >= Ranks[_rankName].minimumInvestment,
+            "Not enough funding for that rank"
+        );
+        S_Donation memory newDonation = S_Donation(
+            msg.value,
+            _rankName,
+            _claimReward
+        );
+        Total += msg.value;
+        Donations[msg.sender].donations.push(newDonation);
+        Donations[msg.sender].total += msg.value;
+    }
+
+    event fundingRetrieved(uint256 _total);
+
+    function retrieveFunding() public onlyOwner {
+        require(block.timestamp >= endDate, "Funding not ended");
+        require(Total >= goal, "Goal not achieved");
+        payable(owner()).transfer(Total);
+        emit fundingRetrieved(Total);
+    }
+
+    function requestRefundGoalNotCompleted() public {
+        require(block.timestamp >= endDate, "Funding not ended");
+        require(Total < goal, "Goal achieved");
+        require(
+            Donations[msg.sender].total > 0,
+            "You didn't fund this project"
+        );
+        payable(msg.sender).transfer(Donations[msg.sender].total);
```

Comment:
user can withdraw multiple times!
Please ensure the remaining user **investment is set to 0**
Also be careful of **reentrancy attacks**